File: backend/internal/hub/crdt/parity_test.go
package crdt_test

import (
	"bytes"
	"fmt"
	"math"
	"math/rand"
	"sort"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"google.golang.org/protobuf/proto"

	leapmuxv1 "github.com/leapmux/leapmux/generated/proto/leapmux/v1"
	"github.com/leapmux/leapmux/internal/hub/crdt"
)

// canonicalizeState produces a deterministic byte encoding for an
// UserCrdtState by sorting maps by key, marshaling each entry, and
// concatenating the marshaled bytes. This is the recipe the plan
// prescribes for parity comparison: protojson is non-deterministic for
// maps, and proto's `Deterministic: true` only orders within a single
// marshal — so we sort and serialize manually.
//
// The exact encoding is:
//
//	[0x01][marshaled NodeRecord per node, sorted by node_id]
//	[0x02][marshaled TabRecord per tab, sorted by tab_id]
//	[0x03][marshaled FloatingWindowRecord per window, sorted by window_id]
//	[0x04][marshaled WorkspaceContentsRecord per workspace, sorted by workspace_id]
//
// Each marshaled record is preceded by its 4-byte big-endian length so
// the consumer can re-walk the stream without ambiguity.
func canonicalizeState(t *testing.T, state *leapmuxv1.UserCrdtState) []byte {
	t.Helper()
	var buf bytes.Buffer
	mopts := proto.MarshalOptions{Deterministic: true}

	keys := make([]string, 0, len(state.GetNodes()))
	for k := range state.GetNodes() {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	buf.WriteByte(0x01)
	for _, k := range keys {
		bs, err := mopts.Marshal(state.GetNodes()[k])
		require.NoError(t, err)
		writeLenPrefixed(&buf, bs)
	}

	keys = keys[:0]
	for k := range state.GetTabs() {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	buf.WriteByte(0x02)
	for _, k := range keys {
		bs, err := mopts.Marshal(state.GetTabs()[k])
		require.NoError(t, err)
		writeLenPrefixed(&buf, bs)
	}

	keys = keys[:0]
	for k := range state.GetFloatingWindows() {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	buf.WriteByte(0x03)
	for _, k := range keys {
		bs, err := mopts.Marshal(state.GetFloatingWindows()[k])
		require.NoError(t, err)
		writeLenPrefixed(&buf, bs)
	}

	keys = keys[:0]
	for k := range state.GetWorkspaces() {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	buf.WriteByte(0x04)
	for _, k := range keys {
		bs, err := mopts.Marshal(state.GetWorkspaces()[k])
		require.NoError(t, err)
		writeLenPrefixed(&buf, bs)
	}
	return buf.Bytes()
}

func writeLenPrefixed(b *bytes.Buffer, data []byte) {
	n := uint32(len(data))
	b.WriteByte(byte(n >> 24))
	b.WriteByte(byte(n >> 16))
	b.WriteByte(byte(n >> 8))
	b.WriteByte(byte(n))
	b.Write(data)
}

// applyAllParity applies a list of validated, canonical-HLC-stamped
// ops to a fresh state with a workspace seeded. Distinct from
// commute_test.go's `applyAll` because the parity tests require a
// preseeded WorkspaceContentsRecord so node-tree projection is
// observable.
func applyAllParity(ops []*leapmuxv1.CrdtOp) *leapmuxv1.UserCrdtState {
	state := crdt.NewState("user-1")
	state.Workspaces["w1"] = &leapmuxv1.WorkspaceContentsRecord{WorkspaceId: "w1", RootNodeId: "root1"}
	for _, op := range ops {
		crdt.Apply(state, op)
	}
	return state
}

// shuffledParity returns a permutation of ops driven by the supplied rng.
func shuffledParity(ops []*leapmuxv1.CrdtOp, rng *rand.Rand) []*leapmuxv1.CrdtOp {
	out := make([]*leapmuxv1.CrdtOp, len(ops))
	copy(out, ops)
	rng.Shuffle(len(out), func(i, j int) { out[i], out[j] = out[j], out[i] })
	return out
}

// TestParity_ManyPermutationsConverge runs several hundred random
// permutations of a 12-op log and asserts they all canonicalize to
// the same bytes. This is the by-construction property the plan calls
// "byte-equal post-state for any permutation of a validated committed
// log".
func TestParity_ManyPermutationsConverge(t *testing.T) {
	// Build a 12-op log: a couple of tabs + node mutations + tombstones.
	ops := []*leapmuxv1.CrdtOp{
		stamped(&leapmuxv1.SetNodeRegisterOp{
			NodeId: "root1",
			Field:  &leapmuxv1.SetNodeRegisterOp_Kind{Kind: leapmuxv1.NodeKind_NODE_KIND_LEAF},
		}, hlcAt(1, 0, "a")),
		stamped(&leapmuxv1.SetTabRegisterOp{
			TabType: leapmuxv1.TabType_TAB_TYPE_AGENT, TabId: "tA",
			Field: &leapmuxv1.SetTabRegisterOp_TileId{TileId: "root1"},
		}, hlcAt(2, 0, "a")),
		stamped(&leapmuxv1.SetTabRegisterOp{
			TabType: leapmuxv1.TabType_TAB_TYPE_AGENT, TabId: "tA",
			Field: &leapmuxv1.SetTabRegisterOp_WorkerId{WorkerId: "w1"},
		}, hlcAt(2, 1, "a")),
		stamped(&leapmuxv1.SetTabRegisterOp{
			TabType: leapmuxv1.TabType_TAB_TYPE_AGENT, TabId: "tA",
			Field: &leapmuxv1.SetTabRegisterOp_Position{Position: "p1"},
		}, hlcAt(2, 2, "a")),
		// Concurrent client b: opens a different tab.
		stamped(&leapmuxv1.SetTabRegisterOp{
			TabType: leapmuxv1.TabType_TAB_TYPE_TERMINAL, TabId: "tB",
			Field: &leapmuxv1.SetTabRegisterOp_TileId{TileId: "root1"},
		}, hlcAt(3, 0, "b")),
		stamped(&leapmuxv1.SetTabRegisterOp{
			TabType: leapmuxv1.TabType_TAB_TYPE_TERMINAL, TabId: "tB",
			Field: &leapmuxv1.SetTabRegisterOp_WorkerId{WorkerId: "w1"},
		}, hlcAt(3, 1, "b")),
		stamped(&leapmuxv1.SetTabRegisterOp{
			TabType: leapmuxv1.TabType_TAB_TYPE_TERMINAL, TabId: "tB",
			Field: &leapmuxv1.SetTabRegisterOp_Position{Position: "p2"},
		}, hlcAt(3, 2, "b")),
		// Two concurrent ratio updates (different clients).
		stamped(&leapmuxv1.SetNodeRegisterOp{
			NodeId: "root1",
			Field: &leapmuxv1.SetNodeRegisterOp_Ratios{Ratios: &leapmuxv1.DoubleList{
				Values: []float64{0.6, 0.4},
			}},
		}, hlcAt(4, 0, "a")),
		stamped(&leapmuxv1.SetNodeRegisterOp{
			NodeId: "root1",
			Field: &leapmuxv1.SetNodeRegisterOp_Ratios{Ratios: &leapmuxv1.DoubleList{
				Values: []float64{0.3, 0.7},
			}},
		}, hlcAt(5, 0, "b")),
		// Tombstone tA at higher HLC (remove-wins).
		stamped(&leapmuxv1.TombstoneTabOp{
			TabType: leapmuxv1.TabType_TAB_TYPE_AGENT, TabId: "tA",
		}, hlcAt(10, 0, "a")),
		// Late SetTab on tA (after tombstone) — must be dropped.
		stamped(&leapmuxv1.SetTabRegisterOp{
			TabType: leapmuxv1.TabType_TAB_TYPE_AGENT, TabId: "tA",
			Field: &leapmuxv1.SetTabRegisterOp_Position{Position: "late"},
		}, hlcAt(11, 0, "a")),
		// Floating-window double register with -0.0 → +0.0 normalization.
		stamped(&leapmuxv1.SetFloatingWindowRegisterOp{
			WindowId: "fw1",
			Field:    &leapmuxv1.SetFloatingWindowRegisterOp_Opacity{Opacity: math.Copysign(0, -1)},
		}, hlcAt(12, 0, "a")),
	}

	canonical := canonicalizeState(t, applyAllParity(ops))

	// 200 random permutations should all canonicalize to the same bytes.
	rng := rand.New(rand.NewSource(0x1234))
	for i := 0; i < 200; i++ {
		state := applyAllParity(shuffledParity(ops, rng))
		got := canonicalizeState(t, state)
		if !bytes.Equal(canonical, got) {
			t.Fatalf("permutation %d produced different bytes (len=%d vs %d)", i, len(canonical), len(got))
		}
		// Convergence alone doesn't say the projection is sane. Check the
		// rendered-tab invariant on every permuted state too: it costs nothing
		// here and gives the property hundreds of orderings to break on.
		assertRenderedTabsAddressable(t, state, fmt.Sprintf("permutation %d", i))
	}
}

// assertRenderedTabsAddressable is the invariant that makes RenderedTabs safe
// to derive a write target from.
//
// `workspace_tab_rendered` backs LocateTab / GetTab / ListTabs, and the CLI
// turns a `--tab-id` into a `--tile-id` through it before emitting
// SetTabRegister(tile_id). `validateTabPlacement` accepts only a live LEAF, so
// every rendered tile must BE one — and must be the tab's own tile_id, not a
// relabelled stand-in. Relabelling is not hypothetical: an earlier frontend
// projection remapped a collapsed SPLIT's tabs onto the parent's node id, which
// is what made a tab render on no tile at all when a grid was closed beside it.
// Both sides now report the raw tile_id (see the collapse cases in
// testdata/crdt_projection_conformance.json); porting a remap here would break
// `tab open --after-tab` in a way no existing test noticed.
func assertRenderedTabsAddressable(t *testing.T, state *leapmuxv1.UserCrdtState, label string) {
	t.Helper()
	proj := crdt.Project(state)
	for _, tab := range proj.RenderedTabs {
		rec := state.GetTabs()[tab.TabID]
		if rec == nil {
			t.Fatalf("%s: rendered tab %q has no TabRecord", label, tab.TabID)
		}
		if raw := rec.GetTileId().GetValue(); tab.TileID != raw {
			t.Fatalf("%s: rendered tab %q reports tile %q but its tile_id register says %q -- "+
				"a relabelled tile is not writable", label, tab.TabID, tab.TileID, raw)
		}
		node := state.GetNodes()[tab.TileID]
		if node == nil {
			t.Fatalf("%s: rendered tab %q sits on unknown tile %q", label, tab.TabID, tab.TileID)
		}
		if node.GetKind().GetValue() != leapmuxv1.NodeKind_NODE_KIND_LEAF {
			t.Fatalf("%s: rendered tab %q sits on tile %q of kind %v -- only a LEAF accepts a tab",
				label, tab.TabID, tab.TileID, node.GetKind().GetValue())
		}
		if !crdt.HLCIsZero(node.GetTombstoneAt()) {
			t.Fatalf("%s: rendered tab %q sits on tombstoned tile %q", label, tab.TabID, tab.TileID)
		}
	}
}

// TestParity_NegativeZeroCanonicalizationIsByteEqual asserts that two
// inputs differing only in zero-sign produce byte-equal canonical
// states.
func TestParity_NegativeZeroCanonicalizationIsByteEqual(t *testing.T) {
	posOps := []*leapmuxv1.CrdtOp{
		stamped(&leapmuxv1.SetFloatingWindowRegisterOp{
			WindowId: "fw1",
			Field:    &leapmuxv1.SetFloatingWindowRegisterOp_Opacity{Opacity: 0.0},
		}, hlcAt(1, 0, "a")),
	}
	negOps := []*leapmuxv1.CrdtOp{
		stamped(&leapmuxv1.SetFloatingWindowRegisterOp{
			WindowId: "fw1",
			Field:    &leapmuxv1.SetFloatingWindowRegisterOp_Opacity{Opacity: math.Copysign(0, -1)},
		}, hlcAt(1, 0, "a")),
	}
	posCanon := canonicalizeState(t, applyAllParity(posOps))
	negCanon := canonicalizeState(t, applyAllParity(negOps))
	assert.True(t, bytes.Equal(posCanon, negCanon),
		"+0.0 and -0.0 canonical bytes must match (sign-bit normalization)")
}

File: frontend/src/lib/crdt/parity.test.ts
import type { UserCrdtState } from '~/generated/proto/leapmux/v1/user_crdt_pb'
import type { CrdtOp } from '~/generated/proto/leapmux/v1/user_ops_pb'
import { create, toBinary } from '@bufbuild/protobuf'
import { describe, expect, it } from 'vitest'
import {
  FloatingWindowRecordSchema,

  HLCSchema,
  NodeRecordSchema,

  TabRecordSchema,
  WorkspaceContentsRecordSchema,
} from '~/generated/proto/leapmux/v1/user_crdt_pb'
import {

  CrdtOpSchema,
  SetFloatingWindowRegisterOpSchema,
  SetNodeRegisterOpSchema,
  SetTabRegisterOpSchema,
  SetWorkspaceRegisterOpSchema,
  TombstoneTabOpSchema,
  TombstoneWorkspaceOpSchema,
} from '~/generated/proto/leapmux/v1/user_ops_pb'
import { TabType } from '~/generated/proto/leapmux/v1/workspace_pb'
import { applyOp, newState } from './apply'

/**
 * canonicalize follows the same recipe as the Go-side `parity_test.go`
 * canonicalizeState helper: sort each map's keys, marshal each entry
 * with deterministic binary proto, concatenate. The hex output is
 * suitable for byte-equal comparison between independent runs.
 *
 * Covers the workspaces map (section 04) just like the backend helper, so
 * SetWorkspaceRegister / TombstoneWorkspace convergence is observable.
 */
function canonicalize(state: UserCrdtState): string {
  const parts: string[] = []

  parts.push('01:')
  for (const k of Object.keys(state.nodes).sort()) {
    parts.push(`${k}=${bytesToHex(toBinary(NodeRecordSchema, state.nodes[k]))};`)
  }
  parts.push('|02:')
  for (const k of Object.keys(state.tabs).sort()) {
    parts.push(`${k}=${bytesToHex(toBinary(TabRecordSchema, state.tabs[k]))};`)
  }
  parts.push('|03:')
  for (const k of Object.keys(state.floatingWindows).sort()) {
    parts.push(`${k}=${bytesToHex(toBinary(FloatingWindowRecordSchema, state.floatingWindows[k]))};`)
  }
  parts.push('|04:')
  for (const k of Object.keys(state.workspaces).sort()) {
    parts.push(`${k}=${bytesToHex(toBinary(WorkspaceContentsRecordSchema, state.workspaces[k]))};`)
  }
  return parts.join('')
}

function bytesToHex(bytes: Uint8Array): string {
  let out = ''
  for (let i = 0; i < bytes.length; i++) {
    const b = bytes[i].toString(16)
    out += b.length === 1 ? `0${b}` : b
  }
  return out
}

function applyAll(ops: CrdtOp[]): UserCrdtState {
  const state = newState('user')
  for (const op of ops) applyOp(state, op)
  return state
}

function mulberry32(seed: number): () => number {
  let a = seed
  return function () {
    let t = (a += 0x6D2B79F5)
    t = Math.imul(t ^ (t >>> 15), t | 1)
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61)
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296
  }
}

function shuffle<T>(items: T[], seed: number): T[] {
  const rng = mulberry32(seed)
  const out = items.slice()
  for (let i = out.length - 1; i > 0; i--) {
    const j = Math.floor(rng() * (i + 1))
    ;[out[i], out[j]] = [out[j], out[i]]
  }
  return out
}

function hlc(p: bigint, l: bigint, c: string) {
  return create(HLCSchema, { physical: p, logical: l, clientId: c })
}

describe('parity', () => {
  it('many shuffled permutations of a heterogeneous op log converge byte-equal', () => {
    const ops: CrdtOp[] = [
      // Two clients add two tabs concurrently.
      create(CrdtOpSchema, {
        canonicalHlc: hlc(10n, 0n, 'a'),
        body: { case: 'setTabRegister', value: create(SetTabRegisterOpSchema, {
          tabType: TabType.AGENT,
          tabId: 'tA',
          field: { case: 'tileId', value: 'root' },
        }) },
      }),
      create(CrdtOpSchema, {
        canonicalHlc: hlc(10n, 1n, 'a'),
        body: { case: 'setTabRegister', value: create(SetTabRegisterOpSchema, {
          tabType: TabType.AGENT,
          tabId: 'tA',
          field: { case: 'workerId', value: 'w1' },
        }) },
      }),
      create(CrdtOpSchema, {
        canonicalHlc: hlc(11n, 0n, 'b'),
        body: { case: 'setTabRegister', value: create(SetTabRegisterOpSchema, {
          tabType: TabType.TERMINAL,
          tabId: 'tB',
          field: { case: 'tileId', value: 'root' },
        }) },
      }),
      // Tombstone tA at higher HLC.
      create(CrdtOpSchema, {
        canonicalHlc: hlc(50n, 0n, 'a'),
        body: { case: 'tombstoneTab', value: create(TombstoneTabOpSchema, { tabType: TabType.AGENT, tabId: 'tA' }) },
      }),
      // Late SetTab on tA — must drop (remove-wins).
      create(CrdtOpSchema, {
        canonicalHlc: hlc(60n, 0n, 'a'),
        body: { case: 'setTabRegister', value: create(SetTabRegisterOpSchema, {
          tabType: TabType.AGENT,
          tabId: 'tA',
          field: { case: 'tileId', value: 'late' },
        }) },
      }),
      // Floating window with an opacity that includes -0.0 (canonicalized).
      create(CrdtOpSchema, {
        canonicalHlc: hlc(70n, 0n, 'a'),
        body: { case: 'setFloatingWindowRegister', value: create(SetFloatingWindowRegisterOpSchema, {
          windowId: 'fw1',
          field: { case: 'opacity', value: -0.0 },
        }) },
      }),
      // Concurrent ratio writes on a node.
      create(CrdtOpSchema, {
        canonicalHlc: hlc(80n, 0n, 'a'),
        body: { case: 'setNodeRegister', value: create(SetNodeRegisterOpSchema, {
          nodeId: 'split1',
          field: { case: 'ratios', value: { $typeName: 'leapmux.v1.DoubleList', values: [0.6, 0.4] } as never },
        }) },
      }),
      create(CrdtOpSchema, {
        canonicalHlc: hlc(81n, 0n, 'b'),
        body: { case: 'setNodeRegister', value: create(SetNodeRegisterOpSchema, {
          nodeId: 'split1',
          field: { case: 'ratios', value: { $typeName: 'leapmux.v1.DoubleList', values: [0.3, 0.7] } as never },
        }) },
      }),
    ]
    const baseline = canonicalize(applyAll(ops))
    for (let i = 0; i < 100; i++) {
      const got = canonicalize(applyAll(shuffle(ops, i)))
      expect(got).toBe(baseline)
    }
  })

  it('-0.0 produces byte-equal canonical output as +0.0', () => {
    const posOp = create(CrdtOpSchema, {
      canonicalHlc: hlc(10n, 0n, 'a'),
      body: { case: 'setFloatingWindowRegister', value: create(SetFloatingWindowRegisterOpSchema, {
        windowId: 'fw',
        field: { case: 'opacity', value: 0.0 },
      }) },
    })
    const negOp = create(CrdtOpSchema, {
      canonicalHlc: hlc(10n, 0n, 'a'),
      body: { case: 'setFloatingWindowRegister', value: create(SetFloatingWindowRegisterOpSchema, {
        windowId: 'fw',
        field: { case: 'opacity', value: -0.0 },
      }) },
    })
    expect(canonicalize(applyAll([posOp]))).toBe(canonicalize(applyAll([negOp])))
  })

  // Workspace lifecycle create/delete now flow through the op log as
  // SetWorkspaceRegister / TombstoneWorkspace. Workspace IDs are fresh
  // nanoids per CreateWorkspace and soft-deleted IDs are never recycled,
  // so a register and a tombstone for the SAME workspace id never co-occur
  // in a real op log — the commuting case is independent workspaces, which
  // this pins: permuting creates/deletes of distinct workspaces converges.
  it('workspace register/tombstone ops on distinct ids converge under any permutation', () => {
    const ops: CrdtOp[] = [
      create(CrdtOpSchema, {
        canonicalHlc: hlc(10n, 0n, 'hub'),
        body: { case: 'setWorkspaceRegister', value: create(SetWorkspaceRegisterOpSchema, { workspaceId: 'w1' }) },
      }),
      create(CrdtOpSchema, {
        canonicalHlc: hlc(20n, 0n, 'hub'),
        body: { case: 'tombstoneWorkspace', value: create(TombstoneWorkspaceOpSchema, { workspaceId: 'w2' }) },
      }),
      create(CrdtOpSchema, {
        canonicalHlc: hlc(30n, 0n, 'hub'),
        body: { case: 'setWorkspaceRegister', value: create(SetWorkspaceRegisterOpSchema, { workspaceId: 'w3' }) },
      }),
    ]
    const baseline = canonicalize(applyAll(ops))
    for (let i = 0; i < 50; i++) {
      const got = canonicalize(applyAll(shuffle(ops, i)))
      expect(got).toBe(baseline)
    }
    // w1 and w3 seeded, w2 tombstoned (was never created) — final state.
    const final = applyAll(ops).workspaces
    expect(final.w1).toBeDefined()
    expect(final.w2).toBeUndefined()
    expect(final.w3).toBeDefined()
  })
})
